### ISSUE TITLE
Allow ValueOptions to be used as optional parameter types

### ISSUE THREAD

**Issue by Happypig375** [fsharp/fslang-suggestions#1136]
# Allow ValueOptions to be used as optional parameter types

I propose we allow:
```fs
type C(?a) =
    let b = defaultValueArg a 2 // Error here
    member _.B = b
```
```
error FS0001: This expression was expected to have type
    ''a voption'    
but here has type
    ''b option'    
```

The existing way of approaching this problem in F# is using reference options. 

## Pros and Cons 

The advantages of making this adjustment to F# are
1. More structs, less GC pressure
2. Allows for more optimisation
3. In line with recent changes to allow ValueOptions to be used as active pattern return types

The disadvantage of making this adjustment to F# is that existing code `?a` will have another potential type for all of us to learn about.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): S

Related suggestions: all are ValueOption parity features.
- https://github.com/fsharp/fslang-suggestions/issues/612
- https://github.com/fsharp/fslang-suggestions/issues/703
- https://github.com/fsharp/fslang-suggestions/issues/739

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by charlesroddie** [comment]
There should be a way to do this yes. I don't think `?a:int` syntax should be used as that is the syntax for options. (And disambiguation is important not only to minimize ambiguity but also because it is important to be able to write fully-annotated F# code where a reader doesn't need to do type inference or read subsequent lines to understand the meaning of a given line.)

I suggest:

1. `type C(??v:int)` or `type [<StuctOptionalArguments>] C(?v:int)` as syntactic sugar for `type C([Optional; DefaultParemeterValue(ValueNone)]v: int voption)`.
2. Allow `DefaultParemeterValue(ValueNone)`, [matching C#](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments#optional-arguments) by allowing not only constant expressions but also default struct values including ValueNone and [parameterless struct constructors](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/parameterless-struct-constructors#parameter-default-value-new).

**Comment by Tarmil** [comment]
I think it could be argued that it's not too bad for readers to not be able to tell if `?a:int` is an option or a voption, because from a user's point of view, you still call it the same way. But this point is moot, because no amount of type inference can tell even the compiler whether this should be an option or a voption:

```fsharp
type IMyInterface =
    abstract M : ?a:int -> unit
```

So we do need a separate syntax indeed.

> Semi-related: it's curious that `DefaultParemeterValue(ValueNone)` doesn't work.

Well, since the tag for `ValueNone` is 0, it's equal to `Unchecked.defaultof<T option>`. So you don't need `DefaultParameterValue`: `Optional` is enough.

**Comment by dsyme** [comment]
I've marked this approved-in-principle.  It would be great to have it implemented

**Comment by charlesroddie** [comment]
This suggestion is contained in the RFC https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1104-struct-representations.md

**Comment by dsyme** [comment]
Yup, the best proposal in that RFC is really `[<Struct>] ?x : int` I think, seems natural all in all.  I don't think `struct ?x: int` flies.

**Comment by edgarfgp** [comment]
Closing this as it was implemented in https://github.com/dotnet/fsharp/pull/18098
